drivers: udc_dwc2: Add nRF54L20 UDC DWC2 quirks

e-rk opened the pull request:
PR consists of 3 changes:
* Add option to select 8-bit UTMI+ PHY interface
* Add DTS nodes to nRF54L20
* Add quirks for nRF54L20

e-rk committed:
dts: common: nordic: nrf54l20: add USBHS node

Added USBHS device tree node for nRF54L20.

Signed-off-by: Rafał Kuźnia <rafal.kuznia@nordicsemi.no>

jfischer-no commented on drivers/usb/udc/udc_dwc2_vendor_quirks.h:
```diff
@@ -108,7 +108,7 @@ DT_INST_FOREACH_STATUS_OKAY(QUIRK_STM32F4_FSOTG_DEFINE)
 #endif /*DT_HAS_COMPAT_STATUS_OKAY(st_stm32f4_fsotg) */
 
-#if DT_HAS_COMPAT_STATUS_OKAY(nordic_nrf_usbhs)
+#if DT_HAS_COMPAT_STATUS_OKAY(nordic_nrf_usbhs_nrfs)
```
There is no point in renaming it or postfixing it with -nrfs. And there is no reason to break someone else's work with a compatible rename. You will need to communicate with us internally to find a suitable compatible.

tmon-nordic replied:
There is a direct need for solving the compatible naming issue. Both nRF54H20 and nRF54L20 do have USBHS peripheral. Yet, the nRF54H20 requires nrfs to get information about VBUS and nRF54L20 has USBREG peripheral for that. Therefore I think the `nordic,nrf-usbhs-nrfs` and `nordic,nrf-usbhs-usbreg` is probably the cleanest approach.

e-rk replied:
I have no strong opinion, although I think the `-nrfs` suffix makes the difference clear.

jfischer-no replied:
I am not convinced to approve changes that break other people's work for no reason. "nordic, nrf-usbhs" should not be renamed. Also, new compatilbe needs a better name, something like "nordic, nrf-usbhs-fancy-soc-family". (nrfs is some software, usbreg is ambiguous)

e-rk replied:
I have removed the `nordic,nrf-usbhs` change and the nRF54L20 now has `nordic,nrf-usbhs-nrf54l` compatible instead.

jfischer-no commented on drivers/usb/udc/Kconfig.dwc2:
```diff
@@ -48,3 +48,11 @@ config UDC_DWC2_USBHS_VBUS_READY_TIMEOUT
 	  UDC DWC2 USBHS VBUS ready event timeout. If the VBUS is not ready
 	  and the Nordic USBHS controller is used, the udc_enable() is
 	  blocked for this amount of time. Set it to zero to wait forever.
+
+config UDC_DWC2_PREFER_8BIT_UTMI
+	bool "Prefer 8-bit UTMI+ PHY interface width"
+	depends on UDC_DWC2
+	help
+	  UDC DWC2 may support both 8- and 16-bit UTMI+ PHY interface widths.
+	  Enable this config to prefer 8-bit interface. Otherwise 16-bit interface
+	  will be used.
```
I can not see any justification why 8-bit should be preferred. And Kconfig is just the wrong place to configure it.

tmon-nordic replied:
If there is UTMI+ PHY capable of both 8-bit and 16-bit operation connected to DWC2 controller then there is very little need to use 8-bit (clocking is different between the two, so there hardware behavior changes slightly).

However, it is also possible that there is only 8-bit capable UTMI+ PHY connected to DWC2 controller that supports both 8-bit and 16-bit. In such case this is a killer feature (because 16-bit simply won't work).

jfischer-no commented on drivers/usb/udc/udc_dwc2_vendor_quirks.h:
```diff
@@ -308,6 +308,105 @@ DT_INST_FOREACH_STATUS_OKAY(QUIRK_NRF_USBHS_DEFINE)
+	wrapper->ENABLE = USBHS_ENABLE_PHY_Msk | USBHS_ENABLE_CORE_Msk;
+	wrapper->TASKS_START = 1UL;
+
+	/* Enable interrupts */
+	wrapper->INTENSET = 1UL;
+
+	return 0;
+}
+
+static inline int usbhs_disable_core(const struct device *dev)
+{
+	NRF_USBHS_Type *wrapper = USBHS_DT_WRAPPER_REG_ADDR(0);
+
+	/* Disable interrupts */
+	wrapper->INTENCLR = 1UL;
+
+	wrapper->ENABLE = 0UL;
+	wrapper->TASKS_START = 1UL;
+
+	return 0;
+}
+
+static inline int usbhs_irq_clear(const struct device *dev)
+{
+	NRF_USBHS_Type *wrapper = USBHS_DT_WRAPPER_REG_ADDR(0);
+
+	wrapper->EVENTS_CORE = 0UL;
+
+	return 0;
+}
+
+static inline int usbhs_init_caps(const struct device *dev)
+{
+	struct udc_data *data = dev->data;
+
+	data->caps.can_detect_vbus = true;
+	data->caps.hs = true;
+
+	return 0;
+}
```
The existing can be move within `#if DT_HAS_COMPAT_STATUS_OKAY(x) && DT_HAS_COMPAT_STATUS_OKAY(y) ... #endif` and reused.

tmon-nordic replied:
I think you meant `||` instead of `&&`.

e-rk committed:
drivers: udc_dwc2: Add nRF54L20 UDC DWC2 quirks

The nRF54L20 does not have the NRFS layer and the set of quirks will be
different than in nRF54H20. Added new set of quirks for the nRF54L20
SoC.

Signed-off-by: Rafał Kuźnia <rafal.kuznia@nordicsemi.no>

e-rk commented:
@tmon-nordic @jfischer-no could you take a look?

tmon-nordic commented on drivers/usb/udc/udc_dwc2.c:
```diff
@@ -1881,7 +1881,10 @@ static int udc_dwc2_init_controller(const struct device *dev)
-	if (usb_dwc2_get_ghwcfg4_phydatawidth(ghwcfg4)) {
+	if (IS_ENABLED(CONFIG_UDC_DWC2_PREFER_8BIT_UTMI) &&
+	    usb_dwc2_get_ghwcfg4_phydatawidth(ghwcfg4) == USB_DWC2_GHWCFG4_PHYDATAWIDTH_8_16_BIT) {
+		LOG_INF("Selecting 8-bit interface as preference");
```
I think `LOG_DBG("Using 8-bit UTMI+");` would be more appropriate here.

e-rk replied:
Done

e-rk commented:
CI is green. @tmon-nordic @jfischer-no @masz-nordic  could you review?

e-rk commented:
> CI is green. @tmon-nordic @jfischer-no @masz-nordic could you review?

Ping

e-rk commented:
I have removed the commit that introduced `CONFIG_UDC_DWC2_PREFER_8BIT_UTMI`.
